jax/experimental/pjit.py

```diff
@@ -0,0 +1,212 @@
+    args.append(xb.set_sharding_proto(subc, arg,
+                                      get_sharding_proto(c, n, axis_resources, mesh)))
+
+  out_nodes = xla.jaxpr_subcomp(
+      subc, call_jaxpr, backend, axis_env, (),
+      extend_name_stack(name_stack, wrap_name(name, "pjit")), *args)
+  out_axis_resources = out_axis_resources_thunk()
+  out_nodes = [xb.set_sharding_proto(subc, out,
+                                     get_sharding_proto(c, n, axis_resources, mesh))
+               for out, axis_resources in safe_zip(out_nodes, out_axis_resources)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates unpartitioned dimension
+ArrayAxisPartitioning = Union[pxla.MeshAxisName, Tuple[pxla.MeshAxisName, ...], None]
+# None indicates fully replicated array value
+ArrayPartitioning = Optional[Tuple[ArrayAxisPartitioning, ...]]
+
+@lu.cache
+def _pjit_callable(
+    fun: lu.WrappedFun,
+    in_axis_resources: Tuple[ArrayPartitioning, ...],
+    out_axis_resources_thunk: Callable[[], Tuple[ArrayPartitioning, ...]],
+    resource_env,
+    donated_invars,
+    name: str,
+    *in_avals):
+
+  in_axes = [get_array_mapping(axes) for axes in in_axis_resources]
+  out_axes_thunk = lambda: [get_array_mapping(axes)
+                            for axes in out_axis_resources_thunk()]
+  # TODO(skye): allow for using a submesh of physical_mesh
+  return pxla.mesh_callable(fun, name, None, resource_env.physical_mesh,
+                            in_axes, out_axes_thunk, donated_invars,
+                            True, *in_avals, tile_by_mesh_axes=False)
+
+
+def with_sharding_constraint(x, axis_resources):
+  resource_env = maps.thread_resources.env
+  return sharding_constraint_p.bind(x, axis_resources=axis_resources,
+                                    resource_env=resource_env)
+
+def _sharding_constraint_impl(x, axis_resources, resource_env):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-pjit contexts? (e.g. pmap, control flow)
+  raise NotImplementedError(
+      "with_sharding_constraint() should only be called inside pjit()")
+
+def _sharding_constraint_translation_rule(c, x_node, axis_resources, resource_env):
+  mesh = resource_env.physical_mesh
+  return xb.set_sharding_proto(c, x_node,
+                               get_sharding_proto(c, x_node, axis_resources, mesh))
+
+sharding_constraint_p = core.Primitive("sharding_constraint")
+sharding_constraint_p.def_impl(_sharding_constraint_impl)
+sharding_constraint_p.def_abstract_eval(lambda x, **unused: x)
+ad.deflinear2(sharding_constraint_p,
+              lambda ct, _, axis_resources, resource_env: (
+                  with_sharding_constraint(ct, axis_resources),))
+xla.translations[sharding_constraint_p] = _sharding_constraint_translation_rule
```

Comment:
Does it make sense to add this only to the "tpu" translation rules? Or, throw an error from the translation rule for non-TPU

Reply:
Agh, I realize I forgot to copy the TPU check from `sharded_jit`'s impl to `pjit`'s impl. I think a top-level check will be better than a TPU-only translation rule, since this rule will theoretically work if/when XLA supports GPU sharding. I'll add  the impl check next time I touch this code.